src/coreclr/src/vm/jitinterface.cpp

```diff
@@ -5095,6 +5095,22 @@ void CEEInfo::getCallInfo(
         EX_THROW(EEMessageException, (kMissingMethodException, IDS_EE_MISSING_METHOD, W("?")));
     }
 
+    // If this call is for a LDFTN and the target method has the NativeCallableAttribute,
+    // then validate it adheres to the limitations.
+    if ((flags & CORINFO_CALLINFO_LDFTN) && pMD->HasNativeCallableAttribute())
+    {
+        if (!pMD->IsStatic())
+            EX_THROW(EEResourceException, (kNotSupportedException, W("NotSupported_NonStaticMethod")));
```

Comment:
Documentation for NotSupportedException says: "The exception that is thrown when an invoked method is not supported"

Documentation for InvalidProgramException says: "The exception that is thrown when a program contains invalid Microsoft intermediate language (MSIL) or metadata."

This is invalid metadata - feels like InvalidProgram is more appropriate.

```suggestion
            EX_THROW(EEResourceException, (kInvalidProgramException, W("NotSupported_NonStaticMethod")));
```

Reply:
@MichalStrehovsky I was thinking about this after our offline conversation. I don't think the metadata is wrong. The current limitations are largely an implementation detail and there is nothing wrong with the metadata since it is valid/parsable. I would argue that the `InvalidProgramException` should not be thrown if the compiler emits it. This isn't to say the compiler can't emit an invalid program, but then it would be a compiler bug.

Reply:
Right - we'll want the compilers to be aware of `NativeCallable` methods so that they disallow this and other problematic patterns (e.g. a managed call to a `NativeCallable` method).

To me this feels like it's in the same bucket as e.g. `DllImport` - compilers validate that it's not used illegally and if it is, it's a compiler bug.

Placing DllImport on an instance method will also produce an `InvalidProgramException` (if you bypass validating compilers and go directly to IL):

```
Unhandled Exception: System.InvalidProgramException: Error: PInvoke item (field,method) must be Static.
   at Blah.MessageBox(IntPtr a, IntPtr b, IntPtr c, IntPtr d)
   at Program.Main()
```

It's a behavior that we didn't define and the metadata is therefore not just unsupported - it's invalid.

Reply:
>we'll want the compilers to be aware of NativeCallable methods so that they disallow this and other problematic patterns (e.g. a managed call to a NativeCallable method).

Yep. I reached out to @jaredpar about having Roslyn do some work - https://github.com/dotnet/runtime/issues/32462#issuecomment-588508231. I think you have again convinced me that this should be `InvalidProgramException`. Thanks for entertaining my thoughts.